File: 프로그래머스/Level2/조이스틱.py
# https://programmers.co.kr/learn/courses/30/parts/12244

def solution(name):
    name = list(name)
    answer = 0
    idx = 0
    
    while True:
        answer += min(ord(name[idx]) - ord('A'), ord('Z') - ord(name[idx]) + 1)
        name[idx] = 'A'
        if name.count('A') == len(name):
            return answer
        left, right = 1, 1
        for i in range(1, len(name)):
            if name[idx-i] == 'A':
                left += 1
            else:
                break
        for i in range(1, len(name)):
            if name[idx+i] == 'A':
                right += 1
            else:
                break
        if left < right:
            answer += left
            idx -= left
        else:
            answer += right
            idx += right
    return answer

File: 프로그래머스/Level2/조이스틱.js
// https://programmers.co.kr/learn/courses/30/lessons/42860

function solution(name) {
  let answer = 0,
    pos = 0;
  let words = new Array(name.length).fill('A');
  const zASCII = 90;
  const aASCII = 65;
  while (words.reduce((prev, cur) => prev + cur, '') !== name) {
    const nowASCII = name.charCodeAt(pos);
    const tempCnt =
      zASCII - nowASCII + 1 < nowASCII - aASCII ? zASCII - nowASCII + 1 : nowASCII - aASCII;
    answer += tempCnt;
    words[pos] = name[pos];
    if (words.reduce((prev, cur) => prev + cur, '') === name) break;
    let rightPos = pos,
      leftPos = pos;
    let rightCnt = 0,
      leftCnt = 0;
    while (true) {
      rightPos = (rightPos + 1) % name.length;
      rightCnt++;
      if (words[rightPos] !== name[rightPos]) break;
    }
    while (true) {
      leftPos = leftPos - 1 >= 0 ? leftPos - 1 : leftPos - 1 + name.length;
      leftCnt++;
      if (words[leftPos] !== name[leftPos]) break;
    }
    if (rightCnt <= leftCnt) {
      pos = rightPos;
      answer += rightCnt;
    } else {
      pos = leftPos;
      answer += leftCnt;
    }
  }
  return answer;
}
